deepinv/physics/phase_retrieval.py

```diff
@@ -159,3 +161,233 @@ def __init__(
+
+        if shifts is not None:
+            self.shifts = shifts
+            self.n_img = len(shifts)
+        else:
+            self.n_img = n_img
+            self.fov = fov
+            self.shifts = self.generate_shifts(n_img=n_img, fov=fov)
+
+        self.probe = (
+            self.probe / self.get_overlap_img(self.probe, self.shifts).mean().sqrt()
+        )
+
+    def A(self, x, **kwargs):
+        """
+        Applies the forward operator to the input image `x` by shifting the probe,
+        multiplying element-wise, and performing a 2D Fourier transform.
+
+        :param x: Input image tensor.
+        :return: Concatenated Fourier transformed tensors after applying shifted probes.
+        """
+        op_fft2 = partial(torch.fft.fft2, norm="ortho")
+        f = lambda x, x_shift, y_shift: op_fft2(
+            self.probe * self.shift(x, x_shift, y_shift)
+        )
+        return torch.cat(
+            [f(x, x_shift, y_shift) for (x_shift, y_shift) in self.shifts], dim=1
+        )
+
+    def A_adjoint(self, y, **kwargs):
+        """
+        Applies the adjoint operator to `y`.
+
+        :param y: Transformed image data tensor of size (batch_size, n_img, height, width).
+        :return: Reconstructed image tensor.
+        """
+        op_ifft2 = partial(torch.fft.ifft2, norm="ortho")
+        g = lambda s, x_shift, y_shift: self.shift(
+            self.probe * op_ifft2(s), -x_shift, -y_shift
+        )
+        for i in range(len(self.shifts)):
+            if i == 0:
+                x = g(y[:, i, :, :].unsqueeze(1), self.shifts[i, 0], self.shifts[i, 1])
+            else:
+                x += g(y[:, i, :, :].unsqueeze(1), self.shifts[i, 0], self.shifts[i, 1])
+        return x
+
+    def construct_probe(self, type="disk", probe_radius=10):
+        """
+        Constructs the probe based on the specified type and radius.
+
+        :param str type: Type of probe shape, e.g., "disk".
+        :param int probe_radius: Radius of the probe shape.
+        :return: Tensor representing the constructed probe.
+        """
+        if type == "disk" or type is None:
+            x = torch.arange(self.img_size[0], dtype=torch.float64)
+            y = torch.arange(self.img_size[1], dtype=torch.float64)
+            X, Y = torch.meshgrid(x, y, indexing="ij")
+            probe = torch.zeros(self.img_size, device=self.device)
+            probe[
+                torch.sqrt(
+                    (X - self.img_size[0] // 2) ** 2 + (Y - self.img_size[1] // 2) ** 2
+                )
+                < probe_radius
+            ] = 1
+        else:
+            raise NotImplementedError(f"Probe type {type} not implemented")
+        return probe
+
+    def generate_shifts(self, n_img, fov=None):
+        """
+        Generates the array of probe shifts across the image, based on probe radius and field of view.
+
+        :param size: Size of the image.
+        :param n_img: Number of shifts (must be a perfect square).
```

Comment:
```suggestion
        :param int n_img: Number of shifts (must be a perfect square).
```
The other params `size`, `probe_radius` seem unused?